lib/vmpooler/metrics/promstats.rb

```diff
@@ -0,0 +1,307 @@
+          param_labels: %i[poolname]
+        },
+        connection_available: {
+          mtype: M_GAUGE,
+          docstring: 'vmpooler Redis Connections Available',
+          prom_metric_prefix: "#{@metrics_prefix}_connection_available",
+          param_labels: %i[type provider]
+        },
+        time_to_ready_state: {
+          mtype: M_HISTOGRAM,
+          buckets: POOLER_TIME_BUCKETS,
+          docstring: 'Time taken for machine to read ready state for pool',
+          prom_metric_prefix: "#{@metrics_prefix}_time_to_ready_state",
+          param_labels: %i[poolname]
+        },
+        migrate: {
+          mtype: M_HISTOGRAM,
+          buckets: POOLER_TIME_BUCKETS,
+          docstring: 'vmpooler Time taken to migrate machine for pool',
+          prom_metric_prefix: "#{@metrics_prefix}_migrate",
+          param_labels: %i[poolname]
+        },
+        clone: {
+          mtype: M_HISTOGRAM,
+          buckets: POOLER_TIME_BUCKETS,
+          docstring: 'vmpooler Time taken to Clone Machine',
+          prom_metric_prefix: "#{@metrics_prefix}_clone",
+          param_labels: %i[poolname]
+        },
+        destroy: {
+          mtype: M_HISTOGRAM,
+          buckets: POOLER_TIME_BUCKETS,
+          docstring: 'vmpooler Time taken to Destroy Machine',
+          prom_metric_prefix: "#{@metrics_prefix}_destroy",
+          param_labels: %i[poolname]
+        },
+        connection_waited: {
+          mtype: M_HISTOGRAM,
+          buckets: REDIS_CONNECT_BUCKETS,
+          docstring: 'vmpooler Redis Connection Wait Time',
+          prom_metric_prefix: "#{@metrics_prefix}_connection_waited",
+          param_labels: %i[type provider]
+        }
+      }
+    end
+
+    # Helper to add individual prom metric.
+    # Allow Histograms to specify the bucket size.
+    def add_prometheus_metric(metric_spec, name, docstring)
+      case metric_spec[:mtype]
+      when M_COUNTER
+        metric_class = Prometheus::Client::Counter
+      when M_GAUGE
+        metric_class = Prometheus::Client::Gauge
+      when M_SUMMARY
+        metric_class = Prometheus::Client::Summary
+      when M_HISTOGRAM
+        metric_class = Prometheus::Client::Histogram
+      else
+        raise("Unable to register metric #{name} with metric type #{metric_spec[:mtype]}")
+      end
+
+      if (metric_spec[:mtype] == M_HISTOGRAM) && (metric_spec.key? :buckets)
+        prom_metric = metric_class.new(
+          name.to_sym,
+          docstring: docstring,
+          labels: metric_spec[:param_labels] + [:vmpooler_instance],
+          buckets: metric_spec[:buckets],
+          preset_labels: { vmpooler_instance: @prefix }
+        )
+      else
+        prom_metric = metric_class.new(
+          name.to_sym,
+          docstring: docstring,
+          labels: metric_spec[:param_labels] + [:vmpooler_instance],
+          preset_labels: { vmpooler_instance: @prefix }
+        )
+      end
+      @prometheus.register(prom_metric)
+    end
```

Comment:
Could you simplify this by just pasing `nil` to the bucket param? Not sure if that would work but, if so, it would be cleaner.

Reply:
@genebean Didn't get to this - will look Monday.